disttask/framework/scheduler/manager.go

```diff
@@ -0,0 +1,311 @@
+		if _, ok := m.subtaskExecutorPools[task.Type]; !ok {
+			logutil.Logger(m.logCtx).Error("unknown task type", zap.String("type", task.Type))
+			continue
+		}
+		exist, err := m.subtaskTable.HasSubtasksInStates(m.id, task.ID, proto.TaskStatePending, proto.TaskStateRevertPending)
+		if err != nil {
+			m.onError(err)
+			continue
+		}
+		if !exist {
+			continue
+		}
+		logutil.Logger(m.logCtx).Info("detect new subtask", zap.Any("id", task.ID))
+		m.addHandlingTask(task.ID)
+		t := task
+		err = m.schedulerPool.Run(func() {
+			m.onRunnableTask(ctx, t.ID, t.Type)
+			m.removeHandlingTask(t.ID)
+		})
+		// pool closed
+		if err != nil {
+			m.removeHandlingTask(task.ID)
+			m.onError(err)
+			return
+		}
+	}
+}
+
+// onCanceledTasks cancels the running tasks.
+func (m *Manager) onCanceledTasks(_ context.Context, tasks []*proto.Task) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	for _, task := range tasks {
+		logutil.Logger(m.logCtx).Info("onCanceledTasks", zap.Any("id", task.ID))
+		if cancel, ok := m.mu.handlingTasks[task.ID]; ok && cancel != nil {
+			cancel()
+		}
+	}
+}
+
+// cancelAllRunningTasks cancels all running tasks.
+func (m *Manager) cancelAllRunningTasks() {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	for id, cancel := range m.mu.handlingTasks {
+		logutil.Logger(m.logCtx).Info("cancelAllRunningTasks", zap.Any("id", id))
+		if cancel != nil {
+			cancel()
+		}
+	}
+}
+
+// filterAlreadyHandlingTasks filters the tasks that are already handled.
+func (m *Manager) filterAlreadyHandlingTasks(tasks []*proto.Task) []*proto.Task {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	var i int
+	for _, task := range tasks {
+		if _, ok := m.mu.handlingTasks[task.ID]; !ok {
+			tasks[i] = task
+			i++
+		}
+	}
+	return tasks[:i]
+}
+
+// onRunnableTask handles a runnable task.
+func (m *Manager) onRunnableTask(ctx context.Context, taskID int64, taskType string) {
+	logutil.Logger(m.logCtx).Info("onRunnableTask", zap.Any("id", taskID), zap.Any("type", taskType))
+	if _, ok := m.subtaskExecutorPools[taskType]; !ok {
+		m.onError(errors.Errorf("task type %s not found", taskType))
+		return
+	}
+	// runCtx only used in scheduler.Run, cancel in m.fetchAndFastCancelTasks
+	scheduler := newScheduler(ctx, m.id, taskID, m.subtaskTable, m.subtaskExecutorPools[taskType])
+	scheduler.Start()
+	defer scheduler.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(checkTime):
+		}
+		task, err := m.globalTaskTable.GetTaskByID(taskID)
+		if err != nil {
+			m.onError(err)
+			return
+		}
+		if task.State != proto.TaskStateRunning && task.State != proto.TaskStateReverting {
+			logutil.Logger(m.logCtx).Info("onRunnableTask exit", zap.Any("id", taskID), zap.Any("state", task.State))
+			return
+		}
+		if exist, err := m.subtaskTable.HasSubtasksInStates(m.id, task.ID, proto.TaskStatePending, proto.TaskStateRevertPending); err != nil {
+			m.onError(err)
+			return
+		} else if !exist {
```

Comment:
How about the state running? e.g. the instance is restarted. May we can add a TODO?

Reply:
Disscuss heartbeat later